Src/VimWpf/RelativeLineNumbers/LineNumberDrawer.cs

```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Windows;
+using System.Windows.Controls;
+
+namespace Vim.UI.Wpf.RelativeLineNumbers
+{
+    public class LineNumberDrawer
+    {
+        private readonly Canvas _canvas;
+
+        private readonly ILineFormatTracker _formatTracker;
+        private readonly LineNumberVisualStore _store;
+
+        public LineNumberDrawer(Canvas canvas, ILineFormatTracker formatTracker)
+        {
+            _canvas = canvas
+                ?? throw new ArgumentNullException(nameof(canvas));
+
+            _formatTracker = formatTracker
+                ?? throw new ArgumentNullException(nameof(formatTracker));
+
+            _store = new LineNumberVisualStore(formatTracker);
+        }
+
+        public void UpdateLines(IEnumerable<Line> lines)
+        {
+            ProcessReformatRequests();
+
+            _canvas.Children.Clear();
+
+            double width = _canvas.Width;
+            foreach (var numberTargets in lines.GroupBy(x => x.Number))
+            {
+                var visual = _store[numberTargets.Key];
+            
+                visual.ReplaceRenderTargets(numberTargets, width);
+
+                _canvas.Children.Add(visual);
+            }
+        }
+
+        private void ProcessReformatRequests()
+        {
+            if (!_formatTracker.TryClearReformatRequest())
+            {
+                return;
+            }
+
+            _canvas.Background = _formatTracker.Background;
+            _canvas.Margin = new Thickness(_formatTracker.NumberWidth, 0, 0, 0);
+
+            _store.Clear();

```

Comment:
Why is the store cleared on every request here? Are the values just not re-usable between requests?

Reply:
Reformat request means format of the line number text changes (font, size, color, background), so we can not reuse cached rendered line numbers anymore. On the other hand we don't want to invalidate them all at once as it could be a long task and potentially could cause some visible performance drop during reformatting on slower PCs.

Alternative would be to set "dirty" flag for each object and invalidate it as we draw it for the first time after the update. It would make the code a bit more complicated but I don't see any performance benefit from it as we save only on one allocation which is negligible compared to generating new bitmap for a number with the new formatting style. If you wish I could do that though.